File: app/src/main/java/openfoodfacts/github/scrachx/openfood/network/deserializers/IngredientsWrapperDeserializer.java
package openfoodfacts.github.scrachx.openfood.network.deserializers;

import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.databind.DeserializationContext;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.deser.std.StdDeserializer;
import openfoodfacts.github.scrachx.openfood.models.IngredientResponse;
import openfoodfacts.github.scrachx.openfood.models.IngredientsWrapper;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.Map;

/**
 * Custom deserializer for {@link IngredientsWrapper IngredientsWrapper}
 *
 * @author dobriseb 2018-12-21 inspired by AllergensWrapperDeserializer
 */
public class IngredientsWrapperDeserializer extends StdDeserializer<IngredientsWrapper> {
    public IngredientsWrapperDeserializer() {
        super(IngredientsWrapper.class);
    }

    @Override
    public IngredientsWrapper deserialize(JsonParser jp, DeserializationContext ctxt) throws IOException {
        List<IngredientResponse> ingredients = new ArrayList<>();

        JsonNode mainNode = jp.getCodec().readTree(jp);
        Iterator<Map.Entry<String, JsonNode>> mainNodeIterator = mainNode.fields();

        while (mainNodeIterator.hasNext()) {
            Map.Entry<String, JsonNode> subNode = mainNodeIterator.next();
            JsonNode namesNode = subNode.getValue().get(DeserializerHelper.NAMES_KEY);

            if (namesNode != null) {
                Map<String, String> names = DeserializerHelper.extractNames(namesNode);
                List<String> parents = DeserializerHelper.extractChildNodeAsText(subNode, DeserializerHelper.PARENTS_KEY);
                List<String> children = DeserializerHelper.extractChildNodeAsText(subNode, DeserializerHelper.CHILDREN_KEY);
                String wikiData = (Boolean) subNode.getValue().has(DeserializerHelper.WIKIDATA_KEY) ? subNode.getValue().get(DeserializerHelper.WIKIDATA_KEY).toString() : null;
                ingredients.add(new IngredientResponse(subNode.getKey(), names, parents, children, wikiData));
            }
        }

        IngredientsWrapper wrapper = new IngredientsWrapper();
        wrapper.setIngredients(ingredients);

        return wrapper;
    }
}

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/models/entities/ingredient/IngredientsWrapperDeserializer.kt
package openfoodfacts.github.scrachx.openfood.models.entities.ingredient

import com.fasterxml.jackson.core.JsonParser
import com.fasterxml.jackson.databind.DeserializationContext
import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.deser.std.StdDeserializer
import openfoodfacts.github.scrachx.openfood.utils.DeserializerHelper
import openfoodfacts.github.scrachx.openfood.utils.DeserializerHelper.extractChildNodeAsText
import openfoodfacts.github.scrachx.openfood.utils.DeserializerHelper.extractMapFromJsonNode
import java.io.IOException

/**
 * Custom deserializer for [IngredientsWrapper]
 *
 * @author dobriseb 2018-12-21 inspired by AllergensWrapperDeserializer
 */
class IngredientsWrapperDeserializer : StdDeserializer<IngredientsWrapper>(IngredientsWrapper::class.java) {
    @Throws(IOException::class)
    override fun deserialize(jp: JsonParser, ctxt: DeserializationContext): IngredientsWrapper {
        val ingredients = ArrayList<IngredientResponse>()
        jp.codec.readTree<JsonNode>(jp).fields().forEach {
            val namesNode = it.value[DeserializerHelper.NAMES_KEY]
            if (namesNode != null) {
                val names = extractMapFromJsonNode(namesNode)
                val parents = extractChildNodeAsText(it, DeserializerHelper.PARENTS_KEY)
                val children = extractChildNodeAsText(it, DeserializerHelper.CHILDREN_KEY)
                val wikiData = if (it.value.has(DeserializerHelper.WIKIDATA_KEY)) it.value[DeserializerHelper.WIKIDATA_KEY].toString() else null
                ingredients.add(IngredientResponse(it.key!!, names, parents, children, wikiData))
            }
        }
        return IngredientsWrapper(ingredients)
    }
}